pkg/controller/federatedtypeconfig/controller.go

```diff
@@ -66,83 +64,214 @@ type FederatedTypeConfigController struct {
+
+	controller, err := newController(client, config, fedNamespace, clusterNamespace, targetNamespace)
+	if err != nil {
+		return err
+	}
+	if minimizeLatency {
+		controller.minimizeLatency()
+	}
+	glog.Infof("Starting FederatedTypeConfig controller")
+	controller.Run(stopChan)
+	return nil
+}
+
+// newController returns a new controller to manage FederatedTypeConfig objects.
+func newController(client corev1alpha1client.CoreV1alpha1Interface, config *restclient.Config, fedNamespace, clusterNamespace, targetNamespace string) (*Controller, error) {
+	c := &Controller{
+		client:           client,
+		config:           config,
 		fedNamespace:     fedNamespace,
 		clusterNamespace: clusterNamespace,
 		targetNamespace:  targetNamespace,
 		stopChannels:     make(map[string]chan struct{}),
+		reviewDelay:      time.Second * 10,
+		smallDelay:       time.Second * 3,
+		workQueue:        workqueue.New(),
+		backoff:          flowcontrol.NewBackOff(5*time.Second, time.Minute),
+	}
+
+	c.deliverer = util.NewDelayingDeliverer()
+
+	c.store, c.controller = cache.NewInformer(
+		&cache.ListWatch{
+			// Only watch the federation namespace to ensure
+			// restrictive authz can be applied to a namespaced
+			// control plane.
+			ListFunc: func(options metav1.ListOptions) (pkgruntime.Object, error) {
+				return client.FederatedTypeConfigs(fedNamespace).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return client.FederatedTypeConfigs(fedNamespace).Watch(options)
+			},
+		},
+		&corev1a1.FederatedTypeConfig{},
+		util.NoResyncPeriod,
+		util.NewTriggerOnAllChanges(func(obj pkgruntime.Object) {
+			c.deliverObj(obj, 0, false)
+		}),
+	)
+
+	return c, nil
+}
+
+// minimizeLatency reduces delays and timeouts to make the controller more responsive (useful for testing).
+func (c *Controller) minimizeLatency() {
+	c.reviewDelay = 50 * time.Millisecond
+	c.smallDelay = 20 * time.Millisecond
+}
+
+// Run runs the Controller.
+func (c *Controller) Run(stopChan <-chan struct{}) {
+	go c.controller.Run(stopChan)
+	c.deliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
+		c.workQueue.Add(item)
+	})
+
+	// wait for the caches to synchronize before starting the worker
+	if !cache.WaitForCacheSync(stopChan, c.controller.HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for cache to sync"))
+		return
 	}
-	// Ensure launched controllers are shutdown cleanly
+
+	// TODO: Allow multiple workers.
+	go wait.Until(c.worker, time.Second, stopChan)
+
+	util.StartBackoffGC(c.backoff, stopChan)
+
+	// Ensure all goroutines are cleaned up when the stop channel closes
 	go func() {
 		<-stopChan
+		c.workQueue.ShutDown()
+		c.deliverer.Stop()
 		c.shutDown()
 	}()
+}
+
+type reconciliationStatus int
 
-	gc := &controller.GenericController{
-		Name:             name,
-		Reconcile:        c.Reconcile,
-		InformerRegistry: arguments.ControllerManager,
+const (
+	statusAllOK reconciliationStatus = iota
```

Comment:
(no action required) we should externalize these to be re-used in each controller.

Reply:
That is [already done](https://github.com/kubernetes-sigs/federation-v2/blob/master/pkg/controller/util/constants.go#L33), but this controller only needs a subset of the enum so I defined just what was needed here.